### PR TITLE
Automod implementation

### PR DESCRIPTION
**Pull request by xMaxximum**
# Description

Automod implementation

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

@Aiko-IT-Systems/discatsharp

### CODE REVIEW

File: DisCatSharp/Clients/DiscordClient.Dispatch.cs [hunk 1]
```diff
@@ -228,6 +228,11 @@ await this._payloadReceived.InvokeAsync(this, new PayloadReceivedEventArgs(this.
 
 			#endregion
 
+			// TODO: Hook dispatch
```
Comment (Lulalaby) [inline DisCatSharp/Clients/DiscordClient.Dispatch.cs:231]: Dispatch needs to be hooked. Rn the client doesn't fire this events

File: DisCatSharp/Clients/DiscordClient.Events.cs [hunk 2]
```diff
@@ -260,6 +260,48 @@ public sealed partial class DiscordClient
 		remove => this._guildIntegrationsUpdated.Unregister(value);
 	}
 	private AsyncEvent<DiscordClient, GuildIntegrationsUpdateEventArgs> _guildIntegrationsUpdated;
+	#endregion
+
+	#region Automod
```
Comment (Lulalaby) [inline DisCatSharp/Clients/DiscordClient.Events.cs:265]: Good job, sharded client needs this too, look at old commits for ref

File: DisCatSharp/Net/Abstractions/Rest/RestAutomodRuleModifyPayload.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	{
+		public Optional<string> Name { get; set; }
+
+		public Optional<AutomodEventType> EventType { get; set; }
+
+		public Optional<AutomodTriggerType> TriggerType { get; set; }
+
+		public Optional<AutomodTriggerMetadata> TriggerMetadata { get; set; }
+
+		public Optional<IEnumerable<AutomodAction>> Actions { get; set; }
```
Comment (Lulalaby) [inline DisCatSharp/Net/Abstractions/Rest/RestAutomodRuleModifyPayload.cs:39]: Use IEnumerable for payloads, it's safer

File: DisCatSharp/Net/Rest/DiscordApiClient.cs [hunk 4]
```diff
@@ -1280,6 +1280,132 @@ internal async Task UpdateUserVoiceStateAsync(ulong guildId, ulong userId, ulong
...
+	/// <param name="guildId"></param>
+	/// <param name="name"></param>
+	/// <param name="eventType"></param>
+	/// <param name="triggerType"></param>
+	/// <param name="actions"></param>
+	/// <param name="triggerMetadata"></param>
+	/// <param name="enabled"></param>
+	/// <param name="exemptRoles"></param>
+	/// <param name="exemptChannels"></param>
+	internal async Task<AutomodRule> CreateAutomodRuleAsync(ulong guildId, string name, AutomodEventType eventType, AutomodTriggerType triggerType, IEnumerable<AutomodAction> actions,
```
Comment (Lulalaby) [inline DisCatSharp/Net/Rest/DiscordApiClient.cs:1331]: Removed the duplicate and used RestAutomodRuleModifyPayload

File: DisCatSharp/Net/Rest/DiscordApiClient.cs [hunk 5]
```diff
@@ -2032,7 +2158,7 @@ internal async Task<IReadOnlyList<DiscordChannel>> GetGuildChannelsAsync(ulong g
 	/// <param name="sendStartNotification">Whether everyone should be notified about the stage.</param>
 	/// <param name="privacyLevel">The privacy_level.</param>
 	/// <param name="reason">The reason.</param>
-	internal async Task<DiscordStageInstance> CreateStageInstanceAsync(ulong channelId, string topic, bool sendStartNotification, StagePrivacyLevel privacyLevel, string reason)
+	internal async Task<DiscordStageInstance> CreateStageInstanceAsync(ulong channelId, string topic, bool sendStartNotification, StagePrivacyLevel privacyLevel = StagePrivacyLevel.GuildOnly, string reason = null)
```
Comment (Lulalaby) [inline DisCatSharp/Net/Rest/DiscordApiClient.cs:2161]: Ignore that

File: DisCatSharp/Entities/Guild/DiscordGuild.AuditLog.cs [hunk 6]
```diff
@@ -61,11 +61,11 @@ public async Task<IReadOnlyList<DiscordAuditLogEntry>> GetAuditLogsAsync(int? li
 			if (rmn <= 0) break;
 
 			var alr = await this.Discord.ApiClient.GetAuditLogsAsync(this.Id, rmn, null, last == 0 ? null : last, byMember?.Id, (int?)actionType).ConfigureAwait(false);
-			ac = alr.Entries.Count();
+			ac = alr.Entries.Count;
```
Comment (Lulalaby) [inline DisCatSharp/Entities/Guild/DiscordGuild.AuditLog.cs:64]: Ignore that

**Review by Lulalaby — CHANGES_REQUESTED** [review]
Feedback goes brr

**Comment by Lulalaby** [comment]
Not sure rn, did you added audit log reason to the api client?

**Comment by Lulalaby** [comment]
TODO: Implement functions to DiscordGuild

**Comment by xMaxximum** [comment]
> Not sure rn, did you added audit log reason to the api client?

Nope i don't think so

**Comment by Lulalaby** [comment]
> > Not sure rn, did you added audit log reason to the api client?
> 
> 
> 
> Nope i don't think so

Then add to all api methods _string reason = null_ and grab the audit log header method from another call.

File: DisCatSharp/Clients/DiscordClient.Dispatch.cs [hunk 7]
```diff
@@ -1320,6 +1339,87 @@ internal async Task OnStickersUpdatedAsync(IEnumerable<DiscordSticker> newSticke
...
+		var channel = this.InternalGetCachedChannel((ulong)channelId);
+
+		var ea = new AutomodActionExecutedEventArgs(this.ServiceProvider)
+		{
+			Guild = guild,
+			Action = executedAction,
+			RuleId = ruleId,
+			TriggerType = triggerType,
+			Member =  member,
+			Channel = channel,
```
Comment (Lulalaby) [inline DisCatSharp/Clients/DiscordClient.Dispatch.cs:1412]: channel might be absent as well.
i.e. when raid

File: DisCatSharp/Entities/Guild/DiscordGuild.cs [hunk 8]
```diff
@@ -1019,6 +1019,71 @@ public Task<DiscordBan> GetBanAsync(DiscordUser user)
 		}
 	}
 
+	/// <summary>
+	/// Gets all auto mod rules for a guild.
+	/// </summary>
+	/// <param name="guildId">The guild id to get the rules for.</param>
+	/// <returns>A collection of all rules in the guild.</returns>
+	public Task<ReadOnlyCollection<AutomodRule>> GetAutomodRulesAsync(ulong guildId)
+		=> this.Discord.ApiClient.GetAutomodRulesAsync(guildId);
```
Comment (Lulalaby) [inline DisCatSharp/Entities/Guild/DiscordGuild.cs:1028]:
```suggestion
	public Task<ReadOnlyCollection<AutomodRule>> GetAutomodRulesAsync()
		=> this.Discord.ApiClient.GetAutomodRulesAsync(this.Id);
```

File: DisCatSharp/Entities/Guild/DiscordGuild.cs [hunk 9]
```diff
@@ -1019,6 +1019,71 @@ public Task<DiscordBan> GetBanAsync(DiscordUser user)
...
+
+	/// <summary>
+	/// Gets a specific auto mod rule.
+	/// </summary>
+	/// <param name="guildId">The guild id of the rule.</param>
+	/// <param name="ruleId">The rule id to get.</param>
+	/// <returns>The auto mod rule.</returns>
+	public Task<AutomodRule> GetAutomodRuleAsync(ulong guildId, ulong ruleId)
+		=> this.Discord.ApiClient.GetAutomodRuleAsync(guildId, ruleId);
```
Comment (Lulalaby) [inline DisCatSharp/Entities/Guild/DiscordGuild.cs:1037]:
```suggestion
	public Task<AutomodRule> GetAutomodRuleAsync(ulong ruleId)
		=> this.Discord.ApiClient.GetAutomodRuleAsync(this.Id, ruleId);
```

File: DisCatSharp/Entities/Guild/DiscordGuild.cs [hunk 10]
```diff
@@ -1019,6 +1019,71 @@ public Task<DiscordBan> GetBanAsync(DiscordUser user)
...
+	/// <param name="actions">The actions of the rule.</param>
+	/// <param name="triggerMetadata">The meta data of the rule.</param>
+	/// <param name="enabled">Whether this rule is enabled.</param>
+	/// <param name="exemptRoles">The exempt roles of the rule.</param>
+	/// <param name="exemptChannels">The exempt channels of the rule.</param>
+	/// <param name="reason">The reason for this addition</param>
+	/// <returns>The created rule.</returns>
+	public Task<AutomodRule> CreateAutomodRuleAsync(ulong guildId, string name, AutomodEventType eventType, AutomodTriggerType triggerType, IEnumerable<AutomodAction> actions,
+		AutomodTriggerMetadata triggerMetadata = null, bool enabled = false, IEnumerable<ulong> exemptRoles = null, IEnumerable<ulong> exemptChannels = null, string reason = null)
+		=> this.Discord.ApiClient.CreateAutomodRuleAsync(guildId, name, eventType, triggerType, actions, triggerMetadata, enabled, exemptRoles, exemptChannels, reason);
```
Comment (Lulalaby) [inline DisCatSharp/Entities/Guild/DiscordGuild.cs:1055]:
```suggestion
	public Task<AutomodRule> CreateAutomodRuleAsync(string name, AutomodEventType eventType, AutomodTriggerType triggerType, IEnumerable<AutomodAction> actions,
		AutomodTriggerMetadata triggerMetadata = null, bool enabled = false, IEnumerable<ulong> exemptRoles = null, IEnumerable<ulong> exemptChannels = null, string reason = null)
		=> this.Discord.ApiClient.CreateAutomodRuleAsync(this.Id, name, eventType, triggerType, actions, triggerMetadata, enabled, exemptRoles, exemptChannels, reason);
```

File: DisCatSharp/Entities/Guild/DiscordGuild.cs [hunk 11]
```diff
@@ -1019,6 +1019,71 @@ public Task<DiscordBan> GetBanAsync(DiscordUser user)
...
+	/// <param name="actions">The new actions of the rule.</param>
+	/// <param name="enabled">Whether this rule should be enabled.</param>
+	/// <param name="exemptRoles">The new exempt roles of the rule.</param>
+	/// <param name="exemptChannels">The new exempt channels of the rule.</param>
+	/// <param name="reason">The reason for this modification.</param>
+	/// <returns>The updated rule.</returns>
+	public Task<AutomodRule> ModifyAutomodRuleAsync(ulong guildId, ulong ruleId, Optional<string> name, Optional<AutomodEventType> eventType, Optional<AutomodTriggerType> triggerType,
+		Optional<AutomodTriggerMetadata> metadata, Optional<IEnumerable<AutomodAction>> actions, Optional<bool> enabled, Optional<IEnumerable<ulong>> exemptRoles,
+		Optional<IEnumerable<ulong>> exemptChannels, string reason = null)
+		=> this.Discord.ApiClient.ModifyAutomodRuleAsync(guildId, ruleId, name, eventType, triggerType, metadata, actions, enabled, exemptRoles, exemptChannels, reason);
```
Comment (Lulalaby) [inline DisCatSharp/Entities/Guild/DiscordGuild.cs:1075]:
```suggestion
	public Task<AutomodRule> ModifyAutomodRuleAsync(ulong ruleId, Optional<string> name, Optional<AutomodEventType> eventType, Optional<AutomodTriggerType> triggerType,
		Optional<AutomodTriggerMetadata> metadata, Optional<IEnumerable<AutomodAction>> actions, Optional<bool> enabled, Optional<IEnumerable<ulong>> exemptRoles,
		Optional<IEnumerable<ulong>> exemptChannels, string reason = null)
		=> this.Discord.ApiClient.ModifyAutomodRuleAsync(this.Id, ruleId, name, eventType, triggerType, metadata, actions, enabled, exemptRoles, exemptChannels, reason);
```

File: DisCatSharp/Entities/Guild/DiscordGuild.cs [hunk 12]
```diff
@@ -1019,6 +1019,71 @@ public Task<DiscordBan> GetBanAsync(DiscordUser user)
...
+
+	/// <summary>
+	/// Deletes an auto mod rule.
+	/// </summary>
+	/// <param name="guildId">The guild id.</param>
+	/// <param name="ruleId">The rule id.</param>
+	/// <param name="reason">The reason for this deletion.</param>
+	/// <returns>The deleted auto mod rule.</returns>
+	public Task<AutomodRule> DeleteAutomodRuleAsync(ulong guildId, ulong ruleId, string reason = null)
+		=> this.Discord.ApiClient.DeleteAutomodRuleAsync(guildId, ruleId, reason);
```
Comment (Lulalaby) [inline DisCatSharp/Entities/Guild/DiscordGuild.cs:1085]:
```suggestion
	public Task<AutomodRule> DeleteAutomodRuleAsync(ulong ruleId, string reason = null)
		=> this.Discord.ApiClient.DeleteAutomodRuleAsync(this.Id, ruleId, reason);
```

File: DisCatSharp/EventArgs/Guild/Automod/AutomodActionExecutedEventArgs.cs [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+
+		/// <summary>
+		/// The member which caused this event.
+		/// </summary>
+		public DiscordMember Member { get; internal set; }
+
+		/// <summary>
+		/// The channel this event happened in.
+		/// </summary>
+		public DiscordChannel? Channel { get; internal set; }
```
Comment (Lulalaby) [inline DisCatSharp/EventArgs/Guild/Automod/AutomodActionExecutedEventArgs.cs:63]: Add channel id too as fallback if the channel wasn't found in cache

File: DisCatSharp/Net/Rest/DiscordApiClient.cs [hunk 14]
```diff
@@ -1280,6 +1280,174 @@ internal async Task UpdateUserVoiceStateAsync(ulong guildId, ulong userId, ulong
...
+	/// <param name="eventType">The new event type of the rule.</param>
+	/// <param name="triggerType">The new trigger type of the rule.</param>
+	/// <param name="metadata">The new metadata of the rule.</param>
+	/// <param name="actions">The new actions of the rule.</param>
+	/// <param name="enabled">Whether this rule is enabled.</param>
+	/// <param name="exemptRoles">The new exempt roles of the rule.</param>
+	/// <param name="exemptChannels">The new exempt channels of the rule.</param>
+	/// <param name="reason">The reason for this modification.</param>
+	/// <returns>The updated automod rule</returns>
+	internal async Task<AutomodRule> ModifyAutomodRuleAsync(ulong guildId, ulong ruleId, Optional<string> name, Optional<AutomodEventType> eventType, Optional<AutomodTriggerType> triggerType,Optional<AutomodTriggerMetadata> metadata, Optional<IEnumerable<AutomodAction>> actions,
```
Comment (Lulalaby) [inline DisCatSharp/Net/Rest/DiscordApiClient.cs:1386]:
```suggestion
	internal async Task<AutomodRule> ModifyAutomodRuleAsync(ulong guildId, ulong ruleId, Optional<string> name, Optional<AutomodEventType> eventType, Optional<AutomodTriggerType> triggerType, Optional<AutomodTriggerMetadata> metadata, Optional<IEnumerable<AutomodAction>> actions,
```

**Review by Lulalaby — CHANGES_REQUESTED** [review]

File: DisCatSharp/Enums/DiscordEvent.cs [hunk 15]
```diff
@@ -54,7 +54,7 @@ public class EventHandlerAttribute : Attribute { }
 /// <summary>
 /// All events available in <see cref="DiscordClient"/> for use with <see cref="EventAttribute"/>.
 /// </summary>
-public enum DiscordEvent
+public enum DiscordEvent : long
```
Comment (Lulalaby) [inline DisCatSharp/Enums/DiscordEvent.cs:57]:
```suggestion
public enum DiscordEvent
```

**Review by Lulalaby — DISMISSED** [review]

File: DisCatSharp/Entities/Guild/Automod/AutomodAction.cs [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+		/// <summary>
+		/// The additional meta data needed during execution for this specific action type.
+		/// </summary>
+		[JsonProperty("metadata?", NullValueHandling = NullValueHandling.Ignore)]
+		public AutomodActionMetadata? Metadata { get; set; }
+
+		/// <summary>
+		/// Creates a new empty automod action.
+		/// </summary>
+		public AutomodAction() { }
```
Comment (Lulalaby) [inline DisCatSharp/Entities/Guild/Automod/AutomodAction.cs:49]:
```suggestion
		internal AutomodAction() { }
```

File: DisCatSharp/Entities/Guild/Automod/AutomodTriggerMetadata.cs [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+		/// The internally predefined word-sets which will be searched for in content.
+		/// </summary>
+		[JsonProperty("presets", NullValueHandling = NullValueHandling.Ignore)]
+		public ReadOnlyCollection<AutomodKeywordPresetType>? Presets { get; set; }
+
+		/// <summary>
+		/// The substrings which will be exempt from triggering the preset type.
+		/// </summary>
+		[JsonProperty("allow_list", NullValueHandling = NullValueHandling.Ignore)]
+		public ReadOnlyCollection<string>? AllowList { get; set; }
```
Comment (Lulalaby) [inline DisCatSharp/Entities/Guild/Automod/AutomodTriggerMetadata.cs:52]:
```suggestion
		public List<string>? AllowList { get; set; }
```
set must be list then

**Review by JBraunsmaJr — APPROVED** [review]
LGTM after the suggestions are implemented

File: DisCatSharp/Entities/Guild/Automod/AutomodRule.cs [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+		/// </summary>
+		[JsonProperty("exempt_channels", NullValueHandling = NullValueHandling.Ignore)]
+		public IReadOnlyList<ulong>? ExemptChannels { get; internal set; }
+
+		[JsonIgnore]
+		public DiscordGuild Guild
+			=> this.Discord.Guilds.TryGetValue(this.GuildId, out var guild) ? guild : null;
+
+		/// <summary>
+		/// Modifies this.GuildId, this.Id auto mod rule.
```
Comment (Lulalaby) [inline DisCatSharp/Entities/Guild/Automod/AutomodRule.cs:107]:
```suggestion
		/// Modifies this auto mod rule.
```

**Review by Lulalaby — APPROVED** [review]

**Pull request merged**